Feature/1.6/ipc/multiplex reconnect

lazan opened the pull request:
I think it's best to review commit by commit. I've tried to keep stuff together as best I can... 

It might be good to ignore ws change in some commit.

_"main reviewer:"_  @ahhud

lazan committed:
common - minor refactoring - mutated and added a few things.

* using string::Argument to be able to take std::string_view
* some ns collapses
* ostream operator -> description
* fixed some stuff in the global `.gitignore`
* generalized ipc::receive a bit more to device::receive
* communication::instance::lookup::request -> send instance lookup
  request. caller needs to handle the future reply "manually.

lazan committed:
cli - generalized --state option and the serialization to stdout

Lifted all (?) _admin::cli_ `--state` options to use the generic one.
This will help us to make the state more _human readable_ in future
patches.

lazan committed:
common - add casual specific assert -> precondition validation

This will help us "move" to a more assert based validation for
preconditions. That is, bugs that is based on internal coding bugs
within casual.

lazan committed:
discovery - simplify order unique service/queue message state

Removed the _ordered set_ abstraction and just keep the state in plain
vectors instead.

* The mutator has the responsibility to preserve the invariant
* When the message is handled, an assertion is done to verify the
  invariant (ordered-unique)

lazan committed:
communication: make ipc::send::Coordinator "reconnect" on unavailable

This patch make sure ipc::send::Coordinator tries to "reconnect" the
outbound-device-instance if one is used as a destination.

This will only work if there are no pending messages to the current ipc
destination. But these will in practice fail and handled via the error
callback. These messages does make sense to send to the new destination
anyway (one can argue...). Hence, it will self heal over time (at least
regarding destination lookup).

Fixes #111

lazan committed:
unittest: fix boot order dependencies in some unittests

This adds boot order to mitigate "race conditions" during shutdown.
Instance  lookup for service-manager could happen after it has been shut
downed... Caller waits for ever.

One could argue that this should work any way. At least to respond with
"an error" if DM is in "shutdown  mode".

ahhud commented on middleware/common/include/common/communication/ipc/send.h:
```diff
@@ -157,12 +158,67 @@ namespace casual
+            }
+            else
+            {
+               while( true)
```
There is not a risk for a never ending loop here?

lazan replied:
It's to keep going if we've get multiple interups... (the reconnect throws if fail) The whole exception paradigm does not work well when the "errors" are expected... I'm going to change at least the whole _device_ API to communicate with _std::expected_ instead. Hast to be a lot better, since a lot of the "errors" are indeed expected...

OlofGullnas commented on .gitignore:
```diff
@@ -11,52 +9,38 @@
-# Tag files produced by Global
-#
+# Tag files produced by Global (what is this?)
 GPATH
```
Googlade. Fick up:
GNU Global is a source code tagging system that works the same way across diverse environments, such as Emacs editor, Vi editor, Less viewer, Bash shell, ...

Detta stämmer också med min minnesbild. Tyckte mig känna igen GTAGS med flera i ett Emacs context.

lazan replied:
Thanks

lazan committed:
communication: never wait on instance lookup if domain is in shutdown

With this patch domain-manager always reply directly for an instance
lookup if the domain is in _shutdown mode_ (or worse). There is no
point in "waiting" on instances that is not likely to ever "spawn",
hence this is a more correct semantic.

lazan commented:
@ahhud Any reservations? I think the last commit fixed the "random shutdown instance lookup deadlock" on unittests.
